temporalcli/commands.workflow_reset.go

```diff
@@ -0,0 +1,205 @@
+	resetType, namespace, wid, rid := c.Type.Value, c.Parent.Namespace, c.WorkflowId, c.RunId
+	wfsvc := cl.WorkflowService()
+	switch resetType {
+	case "LastWorkflowTask":
+		return getLastWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	case "LastContinuedAsNew":
+		return getLastContinueAsNewID(ctx, namespace, wid, rid, wfsvc)
+	case "FirstWorkflowTask":
+		return getFirstWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	default:
+		return "", -1, fmt.Errorf("invalid reset type: %s", resetType)
+	}
+}
+
+// Returns event id of the last completed task or id of the next event after scheduled task.
+func getLastWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace: namespace,
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+
+	for more := true; more; more = len(req.NextPageToken) != 0 {
+		resp, err := wfsvc.GetWorkflowExecutionHistoryReverse(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				break
+			} else if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				// if there is no task completed event, set it to first scheduled event + 1
+				workflowTaskEventID = e.GetEventId() + 1
+			}
+		}
+		req.NextPageToken = resp.NextPageToken
+	}
+	if workflowTaskEventID == 0 {
+		return "", 0, errors.New("unable to find any scheduled or completed task")
+	}
+	return
+}
+
+// Returns id of the first workflow task completed event or if it doesn't exist then id of the event after task scheduled event.
+func getFirstWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryRequest{
+		Namespace: namespace,
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+	for more := true; more; more = len(req.NextPageToken) != 0 {
+		resp, err := wfsvc.GetWorkflowExecutionHistory(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				return resetBaseRunID, workflowTaskEventID, nil
```

Comment:
is it possible that `e.GetEventId()` is 0 and `e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED`? May be we can just break here, just like in the GetLastWorkflowTaskEventID? Then these loops will be exactly the same?

Reply:
As far as I'm aware event IDs are _always_ greater than zero, and even if they began at zero the first WorkflowTaskCompleted event must be preceded by three other events so still could not be zero:
1. WorkflowExecutionStarted              
2. WorkflowTaskScheduled                 
3. WorkflowTaskStarted 
4. **WorkflowTaskCompleted**

Reply:
The logic here is "set workflowTaskEventID to right after the first WorkflowTaskScheduledEvent _but_ override it if we encounter WorkflowTaskCompleted"